Provider: Support module-scoped user-agents via `provider_meta`

drewmullen opened the pull request:
<!--- See what makes a good Pull Request at: https://github.com/hashicorp/terraform-provider-awscc/blob/main/contributing/CONTRIBUTING.md --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request
* The resources and data sources in this provider are generated from the CloudFormation schema, so they can only support the actions that the underlying schema supports. For this reason submitted bugs should be limited to defects in the generation and runtime code of the provider. Customizing behavior of the resource, or noting a gap in behavior are not valid bugs and should be submitted as enhancements to AWS via the CloudFormation Open Coverage Roadmap.

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
### Description

This change enables the AWSCC provider to support the [`provider_meta` block](https://developer.hashicorp.com/terraform/internals/provider-meta). Specifically, modules can use this block to append additional details to the user agent sent in all CRUD operations.

For example, the configuration for a module with a `provider` block already declared in the root might look like the following:

```hcl
terraform {
  required_providers {
    awscc = {
      source  = "hashicorp/awscc"
      version = "~> 1.0"
    }
  }

  # custom provider_meta which will appear appended to user agent
  provider_meta "awscc" {
    user_agent = [
      "jb-demo/0.0.1 (a demo module)",
    ]
  }
}

resource "awscc_iam_role" "module" {
  role_name = "jb-test-ua-module"

  assume_role_policy_document = jsonencode({
    Version = "2012-10-17"
    Statement = [
      {
        Action = ["sts:AssumeRole"]
        Effect = "Allow"
        Principal = {
          Service = "ec2.amazonaws.com"
        }
      }
    ]
  })
}
```

All API calls invoked during CRUD operations will have a user agent string with the custom details from `provider_meta` appended at the end. This can be observed in debug level Terraform logs or in the CloudTrail event history in the AWS console.

`http.user_agent="APN/1.0 HashiCorp/1.0 Terraform/1.14.0-beta1 (+https://www.terraform.io) terraform-provider-awscc/dev (+https://registry.terraform.io/providers/hashicorp/awscc) aws-sdk-go-v2/1.39.0 ua/2.1 os/macos lang/go#1.24.6 md/GOOS#darwin md/GOARCH#arm64 api/cloudcontrol#1.28.4 m/n jb-demo/0.0.1 (a demo module)"`

### Relations

Closes #473
Relates #466

drewmullen committed:
build provider_meta feturea into provider

drewmullen commented:
```shell
$ make build && cp $GOPATH/bin/terraform-provider-awscc ~/.terraform.d/plugins/providermeta/hashicorp/awscc/0.0.1/darwin_amd64/
go install
# github.com/hashicorp/terraform-provider-awscc
./main.go:11:32: cannot use provider.New (type func() tfsdk.ProviderWithProviderMeta) as type func() tfsdk.Provider in argument to tfsdk.Serve
make: *** [build] Error 2
```

gdavison commented on internal/provider/provider.go:
```diff
@@ -23,7 +23,7 @@ const (
 )
 
-func New() tfsdk.Provider {
+func New() tfsdk.ProviderWithProviderMeta {
```
We can still return `tfsdk.Provider` here. As long as `GetMetaSchema` is defined on `AwsCloudControlApiProvider`, it will still fulfill the `tfsdk.ProviderWithProviderMeta` interface

drewmullen committed:
add provider_meta block for experimental testing

drewmullen committed:
structs for useragent

drewmullen committed:
wip

drewmullen committed:
extra providerMeta with ElementAs()

drewmullen committed:
wip

drewmullen commented:
Unable to figure out how to extract the hcl provider_meta set in test `TestAccAWSEC2VPC_metadata`. I believe there is something going on here:

https://github.com/hashicorp/terraform-plugin-framework/blob/5c07e076830a6e13f46b356c4a11439382760a85/internal/proto6server/serve.go#L1236-L1239

I'm pretty sure the function `applyResourceChange` should be able to see the `ProviderMeta` content from the `req *tfprotov6.ApplyResourceChangeRequest` but its not populating there, despite the schema getting there.

Once the data is there, i think we have the functionality in place to unpack and add to the new ctx key for user-agent additions

gdavison committed:
Updates provider metadata deserialization

jar-b committed:
Merge branch 'main' into providermeta

Resolved merge conflicts and refactored the method on the provider
struct to satisfy the WithProviderMeta interface. Smoke tests are now
passing, though validation of the behavior with `provider_meta`
configured on a module is not yet complete. This commit simply brings
the branch up to date with `main` so progress can continue.

```console
% awscc-smoke
TF_ACC=1 go test ./internal/aws/logs -v -count 1 -parallel 20 -run=TestAccAWSLogsLogGroup_\|TestAccAWSLogsLogGroupDataSource_ -timeout 180m

=== CONT  TestAccAWSLogsLogGroup_disappears
2025-09-24T10:43:17.335-0400 [ERROR] sdk.proto: Response contains error diagnostic: tf_rpc=ReadDataSource diagnostic_detail="After attempting to read the data source, the API returned a resource not found error for the id provided. Original Error: couldn't find resource" diagnostic_severity=ERROR diagnostic_summary="AWS Data Source Not Found" tf_req_id=fdc69384-3b91-1658-1dab-ee4553603457 tf_provider_addr=registry.terraform.io/hashicorp/awscc tf_data_source_type=awscc_logs_log_group tf_proto_version=6.10
--- PASS: TestAccAWSLogsLogGroupDataSource_NonExistent (1.83s)
--- PASS: TestAccAWSLogsLogGroup_basic (27.73s)
--- PASS: TestAccAWSLogsLogGroupDataSource_basic (29.87s)
--- PASS: TestAccAWSLogsLogGroup_disappears (34.05s)
--- PASS: TestAccAWSLogsLogGroup_update (42.21s)
PASS
ok      github.com/hashicorp/terraform-provider-awscc/internal/aws/logs 43.514s
```

jar-b committed:
chore(deps): remove `replace` directive for `aws-sdk-go-base`

The changes from this branch were merged into main in August 2022.
https://github.com/hashicorp/aws-sdk-go-base/pull/318

jar-b committed:
internal/types: fix unit tests

jar-b committed:
internal/provider: verify provider interfaces are satisfied at comp time

jar-b committed:
internal/generic: prefer `useragent.Context` to set additional user agent products

jar-b committed:
[wip]internal/acctest: tidy EmptyConfigWithProviderMeta

jar-b committed:
internal/generic: read provider meta in all CRUD ops

jar-b committed:
internal/generic: simplify bootstrap context implementation

jar-b committed:
internal/types: align `UserAgentProduct` field names with aws-sdk-go-base

jar-b committed:
chore: fix `misspell` linter finding

jar-b committed:
internal/provider: align `user_agent` field names with aws-sdk-go-base

jar-b committed:
chore: make docs

jar-b committed:
docs: add `provider_meta` section

jar-b committed:
internal/aws/ec2: first working `_providerMeta` test with VPC resource

```console
% make testacc PKG_NAME=internal/aws/ec2 TESTARGS='-run=TestAccAWSEC2VPC_providerMeta'
TF_ACC=1 go test ./internal/aws/ec2 -v -count 1 -parallel 20 -run=TestAccAWSEC2VPC_providerMeta -timeout 180m

--- PASS: TestAccAWSEC2VPC_providerMeta (36.16s)
PASS
ok      github.com/hashicorp/terraform-provider-awscc/internal/aws/ec2  37.026s
```

jar-b committed:
internal/aws/logs: additional `provider_meta` tests

Includes one additional test with `provider_meta` configured in the root directory, and another configured within a module. Because terraform-plugin-framework does not currently copy module contents when using the `ConfigDirectory` field for a test case, this test is always skipped. Once support for modules is available, we can enable this test.

```console
% make testacc PKG_NAME=internal/aws/logs TESTARGS='-run=TestAccAWSLogsLogGroup_providerMeta'
TF_ACC=1 go test ./internal/aws/logs -v -count 1 -parallel 20 -run=TestAccAWSLogsLogGroup_providerMeta -timeout 180m
=== RUN   TestAccAWSLogsLogGroup_providerMetaWithModule
    log_group_resource_test.go:90: The ConfigDirectory field does not currently support copying modules to the working directory
--- SKIP: TestAccAWSLogsLogGroup_providerMetaWithModule (0.00s)
--- PASS: TestAccAWSLogsLogGroup_providerMeta (25.31s)
PASS
ok      github.com/hashicorp/terraform-provider-awscc/internal/aws/logs 26.313s
```

jar-b commented:
Test results from the latest batch of commits.

```console
% make testacc PKG_NAME=internal/aws/ec2 TESTARGS='-run=TestAccAWSEC2VPC_providerMeta'
TF_ACC=1 go test ./internal/aws/ec2 -v -count 1 -parallel 20 -run=TestAccAWSEC2VPC_providerMeta -timeout 180m

--- PASS: TestAccAWSEC2VPC_providerMeta (36.16s)
PASS
ok      github.com/hashicorp/terraform-provider-awscc/internal/aws/ec2  37.026s
```

```console
% make testacc PKG_NAME=internal/aws/logs TESTARGS='-run=TestAccAWSLogsLogGroup_providerMeta'
TF_ACC=1 go test ./internal/aws/logs -v -count 1 -parallel 20 -run=TestAccAWSLogsLogGroup_providerMeta -timeout 180m
=== RUN   TestAccAWSLogsLogGroup_providerMetaWithModule
    log_group_resource_test.go:90: The ConfigDirectory field does not currently support copying modules to the working directory
--- SKIP: TestAccAWSLogsLogGroup_providerMetaWithModule (0.00s)
--- PASS: TestAccAWSLogsLogGroup_providerMeta (25.31s)
PASS
ok      github.com/hashicorp/terraform-provider-awscc/internal/aws/logs 26.313s
```

I've validated the behavior using modules locally. I was unable to replicate this in an acceptance test due to limitations on what is copied to the working directory when using `ConfigDirectory` in a test step. I'll continue to investigate other potential avenues to test this, but it should not hold up this feature from review and merge.

jar-b committed:
internal/provider: prefer `product_name`, `product_version` for user agent

This naming convention aligns the `provider_meta` schema with the arguments nested inside the existing `user_agent` argument in the provider schema.

```console
% make testacc PKG_NAME=internal/aws/logs TESTARGS='-run=TestAccAWSLogsLogGroup_providerMeta'
TF_ACC=1 go test ./internal/aws/logs -v -count 1 -parallel 20 -run=TestAccAWSLogsLogGroup_providerMeta -timeout 180m

=== RUN   TestAccAWSLogsLogGroup_providerMetaWithModule
    log_group_resource_test.go:90: The ConfigDirectory field does not currently support copying modules to the working directory
--- SKIP: TestAccAWSLogsLogGroup_providerMetaWithModule (0.00s)
--- PASS: TestAccAWSLogsLogGroup_providerMeta (27.26s)
PASS
ok      github.com/hashicorp/terraform-provider-awscc/internal/aws/logs 28.270s
```

```console
% make testacc PKG_NAME=internal/aws/ec2 TESTARGS='-run=TestAccAWSEC2VPC_providerMeta'
TF_ACC=1 go test ./internal/aws/ec2 -v -count 1 -parallel 20 -run=TestAccAWSEC2VPC_providerMeta -timeout 180m

--- PASS: TestAccAWSEC2VPC_providerMeta (38.53s)
PASS
ok      github.com/hashicorp/terraform-provider-awscc/internal/aws/ec2  39.509s
```

jar-b committed:
Merge branch 'main' into providermeta

jar-b committed:
chore: changelog

jar-b committed:
Merge branch 'main' into providermeta

jar-b committed:
Merge branch 'main' into providermeta

jar-b committed:
chore: fix changelog placement after merge

jar-b committed:
Merge branch 'main' into providermeta

jar-b committed:
chore: fixup changelog entry

jar-b committed:
internal/generic: fix post-merge compilation error

jar-b committed:
internal/provider: refactor `provider_meta.user_agent` to list of strings type

This is done to align the schema with the standard provider.

```console
% make testacc PKG_NAME=internal/aws/ec2 TESTARGS='-run=TestAccAWSEC2VPC_providerMeta'
TF_ACC=1 go test ./internal/aws/ec2 -v -count 1 -parallel 20 -run=TestAccAWSEC2VPC_providerMeta -timeout 180m

--- PASS: TestAccAWSEC2VPC_providerMeta (31.37s)
PASS
ok      github.com/hashicorp/terraform-provider-awscc/internal/aws/ec2  32.268s
```

```console
% make testacc PKG_NAME=internal/aws/logs TESTARGS='-run=TestAccAWSLogsLogGroup_providerMeta'
TF_ACC=1 go test ./internal/aws/logs -v -count 1 -parallel 20 -run=TestAccAWSLogsLogGroup_providerMeta -timeout 180m

--- PASS: TestAccAWSLogsLogGroup_providerMeta (24.17s)
PASS
ok      github.com/hashicorp/terraform-provider-awscc/internal/aws/logs 25.032s
```

Inspecting the CloudTrail event history, we can see the `CreateResource`
event contains the expected User-Agent information appended.

```
"userAgent": "APN/1.0 HashiCorp/1.0 Terraform/1.14.0-beta3 (+https://www.terraform.io) terraform-provider-awscc/dev (+https://registry.terraform.io/providers/hashicorp/awscc) aws-sdk-go-v2/1.41.0 ua/2.1 os/macos lang/go#1.24.11 md/GOOS#darwin md/GOARCH#arm64 api/cloudcontrol#1.29.8 m/n test-module/0.0.1 (test comment) second-test-module/0.0.2 (second test comment)",
```

Once the CloudControl API receives the request, the subsequent
calls made to other AWS services by cloud control will _not_ include the
additional User-Agent details. Instead the `userAgent` field will just
display the following:

```
"userAgent": "cloudformation.amazonaws.com",
```

jar-b committed:
internal/generic: tweak check for `user_agent` via `provider_meta`

```console
% make testacc PKG_NAME=internal/aws/ec2 TESTARGS='-run=TestAccAWSEC2VPC_providerMeta'
TF_ACC=1 go test ./internal/aws/ec2 -v -count 1 -parallel 20 -run=TestAccAWSEC2VPC_providerMeta -timeout 180m

--- PASS: TestAccAWSEC2VPC_providerMeta (30.16s)
PASS
ok      github.com/hashicorp/terraform-provider-awscc/internal/aws/ec2  31.214s
```

```console
% make smoke
make: Running smoke tests...
make: NOTE: All tests should pass. Error output for sdk.proto, "Response contains error diagnostic" can be ignored.
TF_ACC=1 go test ./internal/aws/logs -v -count 1 -parallel 3 -run=TestAccAWSLogsLogGroup_\|TestAccAWSLogsLogGroupDataSource_ -timeout 180m

--- PASS: TestAccAWSLogsLogGroup_providerMeta (27.49s)
=== CONT  TestAccAWSLogsLogGroupDataSource_NonExistent
2025-12-11T10:01:30.174-0500 [ERROR] sdk.proto: Response contains error diagnostic: tf_req_id=058bc5d0-c157-9e24-4ac3-5b2eea9a6a18 tf_provider_addr=registry.terraform.io/hashicorp/awscc tf_data_source_type=awscc_logs_log_group tf_rpc=ReadDataSource tf_proto_version=6.10 diagnostic_detail="After attempting to read the data source, the API returned a resource not found error for the id provided. Original Error: couldn't find resource" diagnostic_severity=ERROR diagnostic_summary="AWS Data Source Not Found"
--- PASS: TestAccAWSLogsLogGroupDataSource_NonExistent (1.41s)
=== CONT  TestAccAWSLogsLogGroupDataSource_basic
--- PASS: TestAccAWSLogsLogGroup_basic (30.14s)
=== CONT  TestAccAWSLogsLogGroup_disappears
--- PASS: TestAccAWSLogsLogGroup_update (40.80s)
--- PASS: TestAccAWSLogsLogGroupDataSource_basic (24.79s)
--- PASS: TestAccAWSLogsLogGroup_disappears (29.87s)
PASS
ok      github.com/hashicorp/terraform-provider-awscc/internal/aws/logs 61.019s
```

YakDriver reviewed (dismissed):
LGTM 🎉

jar-b committed:
chore: link to `provider_meta` docs in changelog entry
